File: frontend/src/lib/supabase.ts
/**
 * Cliente de Supabase para el frontend
 * 
 * Este módulo proporciona una instancia configurada del cliente de Supabase
 * para usar en toda la aplicación frontend.
 */

import { createClient, SupabaseClient } from '@supabase/supabase-js';

// Validar que las variables de entorno existan
const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL;
const supabaseAnonKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY;

if (!supabaseUrl || !supabaseAnonKey) {
  throw new Error(
    'Las credenciales de Supabase no están configuradas. ' +
    'Verifica que NEXT_PUBLIC_SUPABASE_URL y NEXT_PUBLIC_SUPABASE_ANON_KEY ' +
    'estén definidas en el archivo .env.local'
  );
}

/**
 * Cliente de Supabase configurado para el navegador
 * 
 * Este cliente usa las credenciales públicas (anon key) y es seguro
 * para usar en el frontend. Las políticas RLS (Row Level Security)
 * deben estar configuradas en Supabase para proteger los datos.
 * 
 * @example
 * ```tsx
 * import { supabase } from '@/lib/supabase';
 * 
 * // Autenticación
 * const { data, error } = await supabase.auth.signInWithPassword({
 *   email: 'user@example.com',
 *   password: 'password123'
 * });
 * 
 * // Consultas a la base de datos
 * const { data: products } = await supabase
 *   .from('products')
 *   .select('*')
 *   .limit(10);
 * ```
 */
export const supabase: SupabaseClient = createClient(supabaseUrl, supabaseAnonKey, {
  auth: {
    // Configuración de persistencia de sesión
    persistSession: true,
    autoRefreshToken: true,
    detectSessionInUrl: true,
    
    // Storage personalizado (opcional)
    storage: typeof window !== 'undefined' ? window.localStorage : undefined,
  },
});

/**
 * Hook para obtener la sesión actual del usuario
 * 
 * @returns {Promise<Session | null>} Sesión actual o null si no hay usuario autenticado
 */
export async function getCurrentSession() {
  const { data: { session } } = await supabase.auth.getSession();
  return session;
}

/**
 * Hook para obtener el usuario actual
 * 
 * @returns {Promise<User | null>} Usuario actual o null si no hay usuario autenticado
 */
export async function getCurrentUser() {
  const { data: { user } } = await supabase.auth.getUser();
  return user;
}

/**
 * Función para cerrar sesión
 * 
 * @returns {Promise<void>}
 */
export async function signOut() {
  const { error } = await supabase.auth.signOut();
  if (error) throw error;
}

/**
 * Función helper para subir archivos a Storage
 * 
 * @param bucket - Nombre del bucket
 * @param path - Ruta del archivo en el bucket
 * @param file - Archivo a subir
 * @returns {Promise<string>} URL pública del archivo subido
 */
export async function uploadFile(bucket: string, path: string, file: File): Promise<string> {
  const { data, error } = await supabase.storage
    .from(bucket)
    .upload(path, file, {
      cacheControl: '3600',
      upsert: false
    });
  
  if (error) throw error;
  
  // Obtener URL pública
  const { data: { publicUrl } } = supabase.storage
    .from(bucket)
    .getPublicUrl(data.path);
  
  return publicUrl;
}

/**
 * Función helper para eliminar archivos de Storage
 * 
 * @param bucket - Nombre del bucket
 * @param paths - Ruta(s) del archivo a eliminar
 */
export async function deleteFile(bucket: string, paths: string | string[]) {
  const pathsArray = Array.isArray(paths) ? paths : [paths];
  const { error } = await supabase.storage
    .from(bucket)
    .remove(pathsArray);
  
  if (error) throw error;
}

// Tipos útiles exportados desde Supabase
export type { User, Session } from '@supabase/supabase-js';

File: backend/app/core/supabase.py
"""
Cliente de Supabase para el backend
"""
from typing import Optional
from supabase import create_client, Client
from app.core.config import settings


class SupabaseClient:
    """Cliente singleton de Supabase"""
    
    _instance: Optional[Client] = None
    
    @classmethod
    def get_client(cls) -> Client:
        """
        Obtiene la instancia única del cliente de Supabase.
        
        Returns:
            Client: Cliente de Supabase configurado
            
        Raises:
            ValueError: Si las credenciales de Supabase no están configuradas
        """
        if cls._instance is None:
            if not settings.SUPABASE_URL or not settings.SUPABASE_KEY:
                raise ValueError(
                    "Las credenciales de Supabase no están configuradas. "
                    "Verifica que SUPABASE_URL y SUPABASE_KEY estén en el archivo .env"
                )
            
            cls._instance = create_client(
                supabase_url=settings.SUPABASE_URL,
                supabase_key=settings.SUPABASE_KEY
            )
        
        return cls._instance
    
    @classmethod
    def get_auth(cls):
        """
        Obtiene el cliente de autenticación de Supabase.
        
        Returns:
            Auth: Cliente de autenticación
        """
        return cls.get_client().auth
    
    @classmethod
    def get_storage(cls):
        """
        Obtiene el cliente de almacenamiento de Supabase.
        
        Returns:
            Storage: Cliente de almacenamiento
        """
        return cls.get_client().storage
    
    @classmethod
    def get_database(cls):
        """
        Obtiene el cliente de base de datos de Supabase (PostgREST).
        
        Returns:
            PostgREST: Cliente de base de datos
        """
        return cls.get_client().postgrest


# Instancia global del cliente
def get_supabase_client() -> Client:
    """
    Función helper para obtener el cliente de Supabase.
    
    Returns:
        Client: Cliente de Supabase configurado
        
    Example:
        ```python
        from app.core.supabase import get_supabase_client
        
        supabase = get_supabase_client()
        response = supabase.table('users').select('*').execute()
        ```
    """
    return SupabaseClient.get_client()


# Función helper para autenticación
def get_supabase_auth():
    """
    Función helper para obtener el cliente de autenticación.
    
    Returns:
        Auth: Cliente de autenticación
        
    Example:
        ```python
        from app.core.supabase import get_supabase_auth
        
        auth = get_supabase_auth()
        session = auth.sign_in_with_password({
            "email": "user@example.com",
            "password": "password123"
        })
        ```
    """
    return SupabaseClient.get_auth()


# Función helper para storage
def get_supabase_storage():
    """
    Función helper para obtener el cliente de almacenamiento.
    
    Returns:
        Storage: Cliente de almacenamiento
        
    Example:
        ```python
        from app.core.supabase import get_supabase_storage
        
        storage = get_supabase_storage()
        storage.from_('product-images').upload('image.jpg', file_data)
        ```
    """
    return SupabaseClient.get_storage()
